### ISSUE TITLE
Tracking issue for Silk.NET 3.0

### ISSUE THREAD

**Issue by Perksey** [Ultz/Silk.NET#209]
# Development
- [ ] [Not Started] WebGL bindings - _unclaimed_
- [ ] [Not Started] WebGPU bindings - _unclaimed_
_More to be added_

# Milestones
_To be added_

Dates are a loose estimate and are subject to change.

**Comment by michael-hawker** [comment]
For clarification. Would WebGL bindings mean that a Silk.NET project could run in the browser over Web Assembly (WASM)?

**Comment by Perksey** [comment]
Yeah exactly, our WebGL and WebGPU bindings are intended to work on Blazor WASM once created to maximise portability of .BET 6 apps.

**Comment by michael-hawker** [comment]
MAUI support would also mean WinUI 3/Uno platform support I believe too, so that'd be cool.

**Comment by Perksey** [comment]
We have included MAUI support in our [work in progress software development plan](https://github.com/dotnet/Silk.NET/blob/784a693501ec4ee14b87ac4b65a39229f185f870/documentation/proposals/Proposal%20-%203.0%20%26%203.X%20Software%20Development%20Plan.md) for 3.0 and will likely be our top priority after WinForms and WPF support, which analytics and discord activity indicate are in massive demand.

This is definitely something we are keen on getting in.

**Comment by IsaacMarovitz** [comment]
Any idea when we'll get Metal bindings?

**Comment by Beyley** [comment]
> Any idea when we'll get Metal bindings?

No ETA for Metal, and its not something the team is working on at all right now, but we welcome any contributions to move us forward

**Comment by Perksey** [comment]
@dotnet/silk-dotnet I've tried to extrapolate all the info from the proposals (as currently approved) into this issue (see the updated description). It makes no assumption on the outcomes of the current SilkTouch or SilkTouchX/SilkX work, mostly just wanted this for tracking.

Please let me know if you have any issues with the above.

**Comment by Perksey** [comment]
Likewise, if there's any other untracked promises for 3.0 we've made or the community wants to be made, please let me know!

**Comment by HurricanKai** [comment]
I think GLFW bindings are arguably < 0, I think we agreed those are a good PoC?

**Comment by Perksey** [comment]
Yes you are right... but do you really want a priority -1? That just feels silly! Unless we're using unsigned integers, then we'll a priority 4,294,967,295...

I expect the SilkTouch Design & Development and GLFW bindings tasks to complete at the same time.

**Comment by HurricanKai** [comment]
I mean it doesn't matter so much, but I'd see GLFW PoC as a separate point - leaving Prio 0 GLFW as an additional item

**Comment by Perksey** [comment]
Okay, done thanks.

**Comment by Perksey** [comment]
I've updated the issue to reflect latest developments. This also has been changed to indicate that we're using SDL3 instead of GLFW as I believe is the consensus on the team, primarily because I've backed down haha

**Comment by curin** [comment]
Unless current plans change, generic math should get bumped up to preview 1

**Comment by Perksey** [comment]
# SilkTouch Notes
- Mods should be able to declare ordering, TransformFunctions currently must be before PrettifyNames because PrettifyNames, despite having a ridiculous amount of code accounting for it, isn't fully compatible with having lots of functions with the same name. This is its own issue that needs to be worked through, but for now the solution is just whacking the TransformFunctions at the end of the pipeline.
    - Is this because of return type conflicts? Should we make this a better experience?

**Comment by uzername** [comment]
please, keep WPF and WinForms integration. Those are needed for real

**Comment by sqeezy** [comment]
I am looking into using silk to replace other c# bindings in a project. Is it worth waiting for 3.0? What is the super rough time frame?

**Comment by Perksey** [comment]
I probably wouldn't wait as we have absolutely no guarantees that 3.0 will even release, let alone a timeframe for that happening. However, we are putting massive amounts of work into 3.0 so it probably won't be too far off, but still likely longer than the time it would take to build an app with 2.0 and migrate to 3.0 later.

We hope to have a good migration story for 3.0 but can't promise it as it is a very breaking release.

**Comment by cyraid** [comment]
Keep it up gents. You got this.

**Comment by michaeldawteck** [comment]
> I probably wouldn't wait as we have absolutely no guarantees that 3.0 will even release, let alone a timeframe for that happening. However, we are putting massive amounts of work into 3.0 so it probably won't be too far off, but still likely longer than the time it would take to build an app with 2.0 and migrate to 3.0 later.
> 
> We hope to have a good migration story for 3.0 but can't promise it as it is a very breaking release.

I don't know about others, but I for one am all for an breaking-but-uncompromised release more so than one that misses out on things due to a need for backward compatibility.
Semantic versioning! That's what Major-digit releases are for. 🤭

**Comment by Perksey** [comment]
Thanks! To clarify where my head is at (this is not representative of the team), I am not backing down on breaking the API - our current API surface is not kind to IDEs and for this and many other reasons it must change in 3.0. The idea is if you have a 3.0 package in your dependency tree that you know about, then you’re opting into 3.0’s API. However, I believe that there’s something we can do about ABI compatibility and this is what I want to research further.

Again this is my view, not that of the team, but ABI compatibility is important now that we have e.g. Uno Platform pulling us in - by being a complete ABI break, 3.0 would be unusable on Uno Platform, which is not what I want. I also want to show that we are dependable for these types of big solutions. So the goal would likely be 3.0 shipping reference assemblies with only the 3.0 API, but the assemblies containing a stub implementation of the 2.X ABI to forward to the 3.0 API. This way if you’re referencing a 3.0 package, your IDE will never pick up on the 2.X APIs, but they’ll be there just in case you pull in a package or an assembly that was built against 2.X. This also provides a nice migration story given that you can keep your code that is yet to be migrated in one project referencing 2.X, and slowly start moving it to another project that references 3.0 while still using code from your 2.X project.

However, a lot of research needs to be done into this to see if it’s even feasible (there are a lot of problems to be solved with abusing reference assemblies in this way) so no promises and again I can’t stress enough that this is _my_ idea and the Silk.NET team may shut it down.

**Comment by michaeldawteck** [comment]
Very interesting. Thank you for explaining and breaking this down!

**Comment by Perksey** [comment]
Big tracker update today! Windowing 3.0 has just landed on the experimental feed, it is in theory implemented on all platforms but we haven't tested Android or iOS yet. Win32 bindings have started work thanks to the excellent @curin (who is also reviewing all of my stuff, the guy never stops). Input's proposal work is now approved as well (thanks to a great many contributions from @domportera!), and this is next to start implementation (likely by me). OpenAL bindings remain the last thing after this that stands before a 3.0 Preview 1, and I would personally love if a community contributor wanted to give it a go but if not this should be an easy thing to do at the tail end of the dev cycle. 

Thank you to the experimental feed early adopters, you are amazing and your input is really valuable!

**Comment by Perksey** [comment]
Yes this is true, the proposal is out-of-date, but I vaguely recall putting some "the team reserves the right to change the reference implementation" get-out wording in there so... I guess it's still correct?

**Comment by Perksey** [comment]
Correction: it's this issue that is linking to an out-of-date proposal. The main branch has a rewritten proposal.

**Comment by Perksey** [comment]
@3xau1o yes and on the 3.0 branch we already are.

**Comment by Thaina** [comment]
@Perksey Is 3.0 branch can use OpenGL ES as WebGL when compiled as wasm?

**Comment by Perksey** [comment]
Not today
